### DISCUSSION TITLE
Merging empty channels with collect() and toList()

### DISCUSSION THREAD

**Discussion by glormph** [nextflow-io/nextflow#2875] [Q&A]
I am a happy user of the following pattern, which collects a number of process outputs and feeds them into a merge process:

```
process proc1 {
output:
tuple val(a), file(b), file(c) into chan1
...
}

chan1
  .toList() // create big list of all output ([a2, b2, c2], [a1, b1, c1], ...])
  .map { it.sort( { a, b -> a[0] <=> b[0] }) } // sort list on value a
  .transpose() // transpose list and output multiple items  [a1, a2, ...], [b2, b2, ], [c1, c2,  ]
  .toList() // one output again, for downstream merge process
  .set { merge_in }
```

This works great, until I ran into the problem where there is no output from `proc1` (e.g. empty sample), which ideally would have to finish the pipeline as usual, just not execute anything downstream of `proc1`. This doesnt work, since the channel has been made non-empty by the two `toList()` operators, and output is now `[[]]`. 

If I instead use `collect()`, it runs fine, no empty lists are made and thus the merge process is not executed. Unfortunately `collect` creates a problem in that it seems to flatten the results to a single list, so the sort step doesnt work. So my question is, am I doing this wrong completely or is there some way to set a channel to empty (or close it) when detecting e.g. this:
```
...
.map { it == [[]] ? empty : it }
or 
.ifEmpty ( close_channel )
```
This strikes me as non-intuitive, but I don't know really. Until then I'm going with the alternative in which I do e.g.  `ifEmpty(['NA', 'NA', 'NA'])`, and detect the problem in downstream processes.

**Comment by bentsherman** [comment]
I didn't realize the `collect` operator automatically flattens its inputs, we'll have to update the docs with a note.

Anyway, the best solution I can think of at the moment is to add a `filter` step after the first `toList`:
```
  .filter { it.size() > 0 }
```
